src/os_posix/os_fallocate.c

```diff
@@ -20,28 +20,29 @@
-                                                                     \
-        pfh = (WT_FILE_HANDLE_POSIX *)file_handle;                   \
-                                                                     \
-        remap = (offset != pfh->mmap_size);                          \
-        if (remap)                                                   \
-            __wt_prepare_remap_resize_file(file_handle, wt_session); \
-                                                                     \
-        WT_SYSCALL_RETRY(op, ret);                                   \
-        if (remap) {                                                 \
-            if (ret == 0)                                            \
-                __wt_remap_resize_file(file_handle, wt_session);     \
-            else {                                                   \
-                __wt_release_without_remap(file_handle);             \
-                WT_RET(ret);                                         \
-            }                                                        \
-        }                                                            \
-        return (0);                                                  \
+#define WT_CALL_FUNCTION(op)                                                   \
+    do {                                                                       \
+        WT_DECL_RET;                                                           \
+        WT_FILE_HANDLE_POSIX *pfh;                                             \
+        bool release, remap;                                                   \
+                                                                               \
+        pfh = (WT_FILE_HANDLE_POSIX *)file_handle;                             \
+                                                                               \
+        release = false;                                                       \
+        remap = (offset != pfh->mmap_size);                                    \
+        if (remap)                                                             \
+            __wt_prepare_remap_resize_file(file_handle, wt_session, &release); \
+                                                                               \
+        WT_SYSCALL_RETRY(op, ret);                                             \
+        if (remap && release) {                                                \
+            if (ret == 0)                                                      \
+                __wt_remap_resize_file(file_handle, wt_session);               \
+            else {                                                             \
+                __wt_release_without_remap(file_handle);                       \
+                WT_RET(ret);                                                   \
+            }                                                                  \
```

Comment:
This seems potentially confusing. What if `remap == true` and `release == false`? Having `remap` set suggests we should call `__wt_remap_resize_file()`, but in this case we don't.

Or to frame it differently, the logic here relies on knowing that `__wt_prepare_remap_resize_file()` doesn't set `release` unless the file is memory mapped. And therefore we can use the fact that `release` isn't set to skip the call to `__wt_remap_resize_file()`.

Maybe add a comment explaining what is happening? Or change the code to match the implied logic: 
```
if (remap) {                
    if (ret == 0)                  
        __wt_remap_resize_file(file_handle, wt_session); 
    else if (release) {      
        __wt_release_without_remap(file_handle);                                 
    } 
}
return (ret);
```

Note: I also eliminated the `WT_RET(ret)`, since it seems superfluous. This might be worth doing even if you go with just commenting what's going on here.

Reply:
It took me a while to grasp the original code as well as @keitharnoldsmith's proposed change.  There are some subtleties here.  I made a truth table to map it out.  Keith's suggested change (KS) behaves differently in these cases:
```
  remap=true    release=false   ret=0
             PR code, does nothing, returns 0
             KS code, calls __wt_remap_resize_file(), returns 0

  remap=true    release=false   ret!=0
             PR code, does nothing, returns 0
             KS code, does nothing, returns ret
```
(the rest of the truth table I believe is the same).

When `remap==true` and `release == false` we know (looking at `__wt_prepare_remap_resize_file`), that  `file_handle->mmap_buf == NULL`.  KS code calls `__wt_remap_resize_file`, which does nothing in the case of the null buffer.  So the PR behavior matches KS when ret==0.

When ret!=0, (and remap is true, release is false), KS returns ret which I think is right.  I think that's the only behavior change between PR and KS.

Summary: the change to do `return (ret)` is needed.  The rest of the KS change doesn't change behavior, but I agree is a little easier to understand.

Reply:
Not entirely - if `release = false` then the running thread never locked out other threads. The issue is that if we call `wt_remap_resize_file` the running code believes it has performed the preparation lock-out in `wt_prepare_remap_resize_file`. In the segfault case I'm fixing, the running thread calls `wt_remap_resize_file` (as in the KS version). Another thread has come in and created a buffer, and this running thread will incorrectly unmap it.

Reply:
Ah, right. So I've just suggested recreating the bug you're trying to fix :-)

I think this conversation shows that it is definitely worth a comment to explain why we aren't calling `__wt_remap_resize_file` even if `remap` is set (i.e., in the `release == false` case).